SG14/slot_map.h

```diff
@@ -304,13 +295,12 @@ class slot_map
         swap(next_available_slot_index_, rhs.next_available_slot_index_);
     }
 
-protected:
     // These accessors are not part of P0661R2 but are "modernized" versions
     // of the protected interface of std::priority_queue, std::stack, etc.
-    constexpr Container<mapped_type>& c() & noexcept { return values_; }
-    constexpr const Container<mapped_type>& c() const& noexcept { return values_; }
-    constexpr Container<mapped_type>&& c() && noexcept { return std::move(values_); }
-    constexpr const Container<mapped_type>&& c() const&& noexcept { return std::move(values_); }
+    constexpr Container<mapped_type>& container() & noexcept { return values_; }
+    constexpr const Container<mapped_type>& container() const& noexcept { return values_; }
+    constexpr Container<mapped_type>&& container() && noexcept { return std::move(values_); }
+    constexpr const Container<mapped_type>&& container() const&& noexcept { return std::move(values_); }
```

Comment:
> Make container access public, it is required for C interop.

This diff seems fine, but I wish there were a bit more explicit rationale given, for posterity. What is a realistic example of something you want to do but can't, with the `protected` interface?

(Full disclosure: I believe nothing in a well-designed class should ever be `protected`. Public seems like the right way to go here, a priori. But I was also trying to remain close-ish to the existing container adaptors, which have a protected _data_ member named `c`. Also, giving completely uninhibited access to the underlying container, such that any client code can `insert` into it arbitrarily, seems a bit _too_ wild-west to me. I wish there were some slightly more scary/verbose way of getting the container than the harmless-looking `sm.container()`.)

Suppose we removed these four overloads and replaced them with a single pilfering overload:

    constexpr Container<mapped_type> container() && noexcept { return std::move(values_); }

Would that meet your requirements in re "C interop"?

Reply:
This has popped up a few times while working with `slot_map`. One good example is pushing data to your gpu in opengl. Which expects a pointer + size. I would be OK with a `data` function, but that seems just as dangerous as `container` in my mind.

I don't think you should force the user to inherit the container for something as simple as reading the data. It is true it's very "wild-west", but isn't that what we love? "Here is a shotgun, go play with it." ;)

Your suggestion does not work, as you do not want to move the data out, just access it. Maybe we could start with a const version first and see if that fulfills most needs? The code that requires this is at work, so I'll check if a const only version would work and get back to you on that.

Reply:
If you just want a `data, size` pair, and you know the underlying container is contiguous, then what's wrong with `&sm.front(), sm.size()`?

Alternatively, I think it would be reasonable to expose a `.data()` method (conditionally, iff the underlying container has a `.data()` method).

But it sounds like for your purposes `sm.empty() ? nullptr : &sm.front()` would be 100% fine as a "pointer to the data."

Reply:
Fair point, I don't remember if there was another case, but that should indeed work.